### PR TITLE
Add upgrade command with some refactor

### PR DESCRIPTION
**Pull request by evalsocket**
# TL;DR
- Upgrade Message 
```bash
# on Darwin if flytectl is installed by brew 
A new release of flytectl is available: v0.2.9-12-g5a0d75a → v0.2.20 
To upgrade, run: brew update && brew upgrade flytectl
https://github.com/flyteorg/flytectl/releases/tag/v0.2.20 

# on Darwin and linux if flytectl install directly binary 
A new release of flytectl is available: v0.2.9-12-g5a0d75a → v0.2.20 
To upgrade, run: flytectl upgrade
https://github.com/flyteorg/flytectl/releases/tag/v0.2.20 

# on windows
A new release of flytectl is available: v0.2.9-12-g5a0d75a → v0.2.20 
https://github.com/flyteorg/flytectl/releases/tag/v0.2.20 
```

- Remove latest version check from version 
```bash
1:59:56 ➜ flytectl  version
{
  "App": "flytectl",
  "Build": "5a0d75a",
  "Version": "v0.2.9-12-g5a0d75a",
  "BuildTime": "2021-07-25 02:00:15.161406 +0530 IST m=+0.000919624"
}
 A new release of flytectl is available: v0.2.9-12-g5a0d75a → v0.2.20 
To upgrade, run: brew update && brew upgrade flytectl
https://github.com/flyteorg/flytectl/releases/tag/v0.2.20 
``` 

- Added Upgrade Command 
```bash
# on Linux and Darwin
2:02:18 ➜ flytectl upgrade
Successfully updated to version v0.2.20

# on Darwin if it's brew install
A new release of flytectl is available: v0.2.9-12-g5a0d75a → v0.2.20 
To upgrade, run: brew update && brew upgrade flytectl
https://github.com/flyteorg/flytectl/releases/tag/v0.2.20 

# on windows
2:02:18 ➜ flytectl upgrade

A new release of flytectl is available: v0.2.9-12-g5a0d75a → v0.2.20 
Flytectl auto upgrade is not available on windows
https://github.com/flyteorg/flytectl/releases/tag/v0.2.20 
```

- Added rollback subcommand Command 
```bash
# on Linux and darwin for manual install using curl
2:02:18 ➜ flytectl upgrade rollback
```

- Added check for disk issue in sandbox
```bash
2:09:54 ➜ flytectl sandbox start
🧑‍🏭 Bootstrapping a brand new flyte cluster... 🔨 🔧
delete existing sandbox cluster [y/n]: y
🐋 pulling docker image cr.flyte.org/flyteorg/flyte-sandbox:dind
{"status":"Pulling from flyteorg/flyte-sandbox","id":"dind"}
{"status":"Digest: sha256:1d9888ae76e435c4377e0cb750d920a8ba1117e9ec4f574569b1aa44759011b9"}
{"status":"Status: Image is up to date for cr.flyte.org/flyteorg/flyte-sandbox:dind"}
🧑‍🏭 booting Flyte-sandbox container
92021-07-24T20:48:34.877053112Z Starting Docker daemon...
%2021-07-24T20:48:41.733980550Z Done.
72021-07-24T20:48:41.734032041Z Starting k3s cluster...
%2021-07-24T20:48:47.901579830Z Done.
err: docker sandbox doesn't have sufficient memory available. Please run docker system prune
```

- Remove wait script from docker sandbox and created k8s client for verifying the deployment 
```bahs
2:09:54 ➜ flytectl sandbox start
🧑‍🏭 Bootstrapping a brand new flyte cluster... 🔨 🔧
delete existing sandbox cluster [y/n]: y
🐋 pulling docker image cr.flyte.org/flyteorg/flyte-sandbox:dind
{"status":"Pulling from flyteorg/flyte-sandbox","id":"dind"}
{"status":"Digest: sha256:1d9888ae76e435c4377e0cb750d920a8ba1117e9ec4f574569b1aa44759011b9"}
{"status":"Status: Image is up to date for cr.flyte.org/flyteorg/flyte-sandbox:dind"}
🧑‍🏭 booting Flyte-sandbox container
92021-07-24T20:50:37.461210692Z Starting Docker daemon...
%2021-07-24T20:50:44.492807321Z Done.
72021-07-24T20:50:44.493131574Z Starting k3s cluster...
%2021-07-24T20:50:50.717199072Z Done.
+------------------------------------+-----------+-----------+
|              SERVICE               |  STATUS   | NAMESPACE |
+------------------------------------+-----------+-----------+
| flytepropeller-6bb9bfd44c-p2q2n    | Running   | flyte     |
+------------------------------------+-----------+-----------+
| flyte-pod-webhook-7747467c66-rnzxl | Running   | flyte     |
+------------------------------------+-----------+-----------+
| postgres-6c64cd5c48-gfgc8          | Running   | flyte     |
+------------------------------------+-----------+-----------+
| datacatalog-7db5796c46-lkj6s       | Running   | flyte     |
+------------------------------------+-----------+-----------+
| flyteconsole-66c5575f79-2ddzf      | Running   | flyte     |
+------------------------------------+-----------+-----------+
| minio-8564dc858c-66v46             | Running   | flyte     |
+------------------------------------+-----------+-----------+
| flyteadmin-66d79f98f8-xkzwt        | Running   | flyte     |
+------------------------------------+-----------+-----------+
| syncresources-27141388-9zwz6       | Succeeded | flyte     |
+------------------------------------+-----------+-----------+
| syncresources-27141389-r7f7t       | Succeeded | flyte     |
+------------------------------------+-----------+-----------+
| syncresources-27141390-7mn6v       | Succeeded | flyte     |
+------------------------------------+-----------+-----------+
```
## Type
- [ ] Bug Fix
- [x] Feature
- [ ] Plugin

## Are all requirements met?

- [ ] Code completed
- [ ] Smoke tested
- [ ] Unit tests added
- [ ] Code documentation added
- [ ] Any pending items have an associated Issue

## Complete description
_How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/1233

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: cmd/upgrade/coverage.temp.txt [hunk 1]
```diff
@@ -0,0 +1,41 @@
+mode: atomic
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/coverage.temp.txt:1]: Remove this file

File: .github/workflows/sandbox.yaml [hunk 2]
```diff
@@ -20,6 +20,11 @@ jobs:
           key: ${{ runner.os }}-go-${{ hashFiles('go.sum') }}
       - name: Build Flytectl binary
         run: make compile
+      - name: Setup env
+        run: |
+          mkdir -p ~/.flyte/k3s
+          touch ~/.flyte/k3s/k3s.yaml
+          chmod 777 ~/.flyte/k3s/k3s.yaml
```
Comment (pmahindrakar-oss) [inline .github/workflows/sandbox.yaml:27]: This requires only readwrite and not execute

File: cmd/sandbox/start.go [hunk 3]
```diff
@@ -86,7 +110,8 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
 		if err.Error() != clierrors.ErrSandboxExists {
 			return nil, err
 		}
-		printExistingSandboxMessage()
+		fmt.Printf("Existing details of your sandbox:")
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:113]: Move this inside the function on L114

File: cmd/sandbox/start.go [hunk 4]
```diff
@@ -110,14 +135,23 @@ func startSandbox(ctx context.Context, cli docker.Docker, reader io.Reader) (*bu
...
-			return nil, err
-		}
-		vol, err := mountVolume(FlyteManifest, GeneratedManifest)
+		isGreater, err := util.IsVersionGreaterThan(sandboxConfig.DefaultConfig.Version, flyteMinimumVersionSupported)
 		if err != nil {
 			return nil, err
 		}
-		volumes = append(volumes, *vol)
+		if !isGreater {
+			logger.Infof(ctx, "version flag only supported after with flyte %s+ release", flyteMinimumVersionSupported)
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:143]: Should we abort instead of continuing with pulling the image and just logging it . The action is just to log ? what does that tell to the user to do

File: cmd/sandbox/start.go [hunk 5]
```diff
@@ -164,34 +216,52 @@ func mountVolume(file, destination string) (*mount.Mount, error) {
...
+		if response.StatusCode != 200 {
+			return fmt.Errorf("someting goes wrong while downloading the flyte release")
+		}
+		data, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return err
+		}
+		if err := util.WriteIntoFile(data, FlyteManifest); err != nil {
+			return err
+		}
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:264]: Create this as part of a downloader utiity which can be reused at multiple places

File: cmd/sandbox/start.go [hunk 6]
```diff
@@ -164,34 +216,52 @@ func mountVolume(file, destination string) (*mount.Mount, error) {
...
+	}
+	if len(manifestURL) > 0 {
+		response, err := http.Get(manifestURL)
+		if err != nil {
+			return err
+		}
+		defer response.Body.Close()
+		if response.StatusCode != 200 {
+			return fmt.Errorf("someting goes wrong while downloading the flyte release")
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:256]: log the statuscode

File: cmd/sandbox/start.go [hunk 7]
```diff
@@ -164,34 +216,52 @@ func mountVolume(file, destination string) (*mount.Mount, error) {
...
+		}
+		data, err := ioutil.ReadAll(response.Body)
+		if err != nil {
+			return err
+		}
+		if err := util.WriteIntoFile(data, FlyteManifest); err != nil {
```
Comment (pmahindrakar-oss) [inline cmd/sandbox/start.go:262]: accept FlyteManifest as argument

File: cmd/upgrade/upgrade.go [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+		if err != nil {
+			return err
+		}
+		return nil
+	}, retry.Attempts(3))
+	if err != nil {
+		return err
+	}
+	return nil
+}
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:213]: Can be part of filteutils

File: cmd/upgrade/upgrade.go [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:232]: Same with this function . can be part of fileutils

File: cmd/upgrade/upgrade.go [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+	} else if goos == "darwin" {
+		fmt.Printf("A new release of flytectl is available: %s → %s \n", stdlibversion.Version, latest)
+		fmt.Println("To upgrade, run: brew update && brew upgrade flytectl")
+		fmt.Printf("https://github.com/flyteorg/flytectl/releases/tag/%s \n", latest)
+		return true
+	}
+	return false
+}
+
+func checkGOOSForRollback(goos string) bool {
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:193]: rename to isRollBackSupported

File: cmd/upgrade/upgrade.go [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+	if goos == "windows" || goos == "darwin" {
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:194]: enum for goos

File: cmd/upgrade/upgrade.go [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+		return err
+	}
+	// Check if it's a rollback
+	if len(args) > 0 {
+		if args[0] == "rollback" {
+			if checkGOOSForRollback(goos) {
+				return nil
+			}
+			// Restore the binary
+			source = backup
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:74]: If the backup doesn't exist , then we should error out

File: cmd/upgrade/upgrade.go [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+		}
+	} else {
+		latest, err := util.GetLatestVersion("flytectl")
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:77]: use AppName constant

File: cmd/upgrade/upgrade.go [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+		if err != nil {
+			return err
+		}
+
+		isGreater, err := util.IsVersionGreaterThan(latest.GetTagName(), stdlibversion.Version)
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:82]: Create a separate versionutils

File: cmd/upgrade/upgrade.go [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+		return nil
+	}
+
+	if cmdUtil.AskForConfirmation(fmt.Sprintf("This action will upgrade flytectl version [%s]. Do you want to continue?", latest.GetTagName()), reader) {
+		arch := runtime.GOARCH
+		if arch == "amd64" {
+			arch = "x86_64"
+		} else if arch == "386" {
+			arch = "i386"
+		}
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:123]: Define a flytectl binary class to define these various architectures and there mappings to the asset Name. Need not  be in this class

File: cmd/upgrade/upgrade.go [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+		for _, v := range latest.Assets {
+			asset := fmt.Sprintf("flytectl_%s_%s.tar.gz", strings.Title(goos), arch)
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:125]: Same here . can be part of a binary class which defines this and functionality on getting the asset name

File: cmd/upgrade/upgrade.go [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+				if err != nil {
+					return err
+				}
+				gzr, err := gzip.NewReader(response.Body)
+				if err != nil {
+					return err
+				}
+				defer gzr.Close()
+
+				tr := tar.NewReader(gzr)
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:137]: Common utility to download and read tar

File: cmd/upgrade/upgrade.go [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+									break
+								}
+								return err
+							}
+						}
+						// Backup the binary
+						if err := copy(ext, backup); err != nil {
+							return err
+						}
+						break
```
Comment (pmahindrakar-oss) [inline cmd/upgrade/upgrade.go:169]: Same here can be a part of common functionality to read a tar .
Provide an option to backup as a flag or name of the backup file in the utility

File: pkg/k8s/k8s.go [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error building kubernetes clientset")
+		}
+		return kubeClient, nil
+	}
+	return Client, nil
+}
+
+func GetCountOfReadyDeployment(ctx context.Context, client appsv1.AppsV1Interface) (int64, error) {
+	deployments, err := client.Deployments("flyte").List(ctx, v1.ListOptions{})
```
Comment (pmahindrakar-oss) [inline pkg/k8s/k8s.go:44]: namespace constant

File: .github/workflows/sandbox.yaml (hunk 2 above)
Comment (evalsocket) [reply .github/workflows/sandbox.yaml:27]: yes

File: cmd/sandbox/start.go (hunk 3 above)
Comment (evalsocket) [reply cmd/sandbox/start.go:113]: no `util.PrintSandboxMessage()` is a common function....It also print message on new cluster

File: cmd/sandbox/start.go [hunk 20]
```diff
@@ -76,6 +81,25 @@ func startSandboxCluster(ctx context.Context, args []string, cmdCtx cmdCore.Comm
 	if reader != nil {
 		docker.WaitForSandbox(reader, docker.SuccessMessage)
 	}
+
+	time.Sleep(5 * time.Second)
```
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go:85]: we will wait unitl k3s cluster is ready

[the remark this answers is not in the document]
Comment (pmahindrakar-oss) [reply cmd/sandbox/start.go:85]: This check doesn't seem deterministic .what if it takes longer than the sleep time.

File: .github/workflows/sandbox.yaml (hunk 2 above)
Comment (pmahindrakar-oss) [reply .github/workflows/sandbox.yaml:27]: Can we Reduce the permissions then

File: cmd/sandbox/start.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (evalsocket) [reply cmd/sandbox/start.go]: cool..I added retry until k8s client is ready

File: pkg/k8s/k8s.go [hunk 21]
```diff
@@ -0,0 +1,69 @@
...
+
+func GetFlyteDeployment(ctx context.Context, client corev1.CoreV1Interface) (*corev1api.PodList, error) {
+	pods, err := client.Pods(flyteNamespace).List(ctx, v1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return pods, nil
+}
+
+func GetNodeTaintStatus(ctx context.Context, client corev1.NodeInterface) (bool, error) {
```
Comment (pmahindrakar-oss) [inline pkg/k8s/k8s.go:52]: isNodeTainted

File: pkg/util/util.go [hunk 22]
```diff
@@ -71,3 +47,29 @@ func IsVersionGreaterThan(version1, version2 string) (bool, error) {
...
+	req, _ := http.NewRequest("GET", url, nil)
+	response, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+	if response.StatusCode != 200 {
+		return nil, fmt.Errorf("someting goes wrong while sending request to %s. Got status code %v", url, response.StatusCode)
+	}
+	return response, nil
+}
```
Comment (pmahindrakar-oss) [inline pkg/util/util.go:75]: This can part of the requests_util which takes in url, method, body and return the response

**Comment by evalsocket** [comment]
@pmahindrakar-oss Regarding `I couldn't find definition for this function. Can you point me to it`....You can check the implementation of package github.com/mouuff/go-rocket-update

Regarding `Where are you setting this Client, isn't this always nil` - In k8s package in used client because of testing. In test cases i will override the test client.

File: pkg/docker/docker_util_test.go [hunk 23]
```diff
@@ -28,7 +29,7 @@ var (
 func setupSandbox() {
 	mockAdminClient := u.MockClient
 	cmdCtx = cmdCore.NewCommandContext(mockAdminClient, u.MockOutStream)
-	_ = util.SetupFlyteDir()
+	_ = os.MkdirAll(f.FilePathJoin(f.UserHomeDir(), ".flyte"), os.ModePerm)
```
Comment (pmahindrakar-oss) [inline pkg/docker/docker_util_test.go:32]: Can we log the error instead of ignoring it. Could be permission issue

**Review by pmahindrakar-oss — APPROVED** [review]
